---
title: 领域驱动设计实践(1)-错误处理
date: 2022-07-11
tags: [领域驱动设计]
categories: 领域驱动设计
---

本文讨论了领域驱动设计实践下的错误处理。

<!--more-->

在软件开发过程中，错误处理一直是个比较麻烦的事情。对于错误处理，我们经常需要回答的问题是：

* 在哪些地方可能出现错误
* 当程序某处发生错误时，是将错误交给调用方处理还是在原处处理（比如直接抛出异常）
* 业务错误和HTTP层的状态码要如何关联
* 错误和异常有什么区别

这里我们先不急着回答这些问题，先来看一张图。下面这张图是领域驱动设计的六边形架构的示意图。

![领域驱动设计的六边形架构](/assets/images/post_imgs/domain_driven_hexagon.png)

下面以银行转账这个场景为例子，假设用户在网上银行要从 A 账户向 B 账户转账 1000 元。但 A 账户只有 800 元，余额不足，这将导致一个账户余额不足错误(Insufficient balance error)。

## 在调用点抛出错误

在调用点抛出异常，伪代码如下：

```
class Account {
    private balance: number;
    
    transfer(amount: number, receiptAccount: Account) {
        if (this.balance < amount) {
            throw new Error("Insufficient balance error");
        }
        ...
    }
}
```

这种方式非常直接，在业务错误点就地抛出一个错误，错误的全部信息都在传入的字符串里。transfer 方法的调用方需要使用 try/catch 包裹这个方法的调用（前提是当前使用的编程语言支持try/catch）。接着在 catch 块中对捕获到的错误进行处理：

```
class TransferBusinessService {
    transfer(amount: number, paymentAccount: Account, receiptAccount: Account) {
        try {
            paymentAccount.transfer(amount, receiptAccount);
        } catch (err) {
            // handle error by error message
        }
    }
}
```

TransferBusinessService 作为调用方，在错误处理时只能利用错误信息里的字符串，除此之外没有其他办法。那么考虑升级一下，下一步我们封装出错误对象来。

## 抛出错误对象

引入一个错误对象 InsufficientBalanceError，在出错时抛出这个错误对象：

```
class InsufficientBalanceError extends Error {
    static readonly message = 'Insufficient balance error';

    public readonly code = 'TRANSFER.INSUFFICIENT_BALANCE_ERROR';

    constructor(metadata?: unknown) {
        super(InsufficientBalanceError.message, metadata);
    }
}
```

```
class Account {
    private balance: number;
    
    transfer(amount: number, receiptAccount: Account) {
        if (this.balance < amount) {
            throw InsufficientBalanceError();
        }
        ...
    }
}
```

调用方代码：

```
class TransferBusinessService {
    transfer(amount: number, paymentAccount: Account, receiptAccount: Account) {
        try {
            paymentAccount.transfer(amount, receiptAccount);
        } catch (err) {
             // handle error by error type
        }
    }
}
```

在调用方的错误处理中，从之前的基于错误消息字符串变成了基于错误类型。已经具备了初步的错误建模能力。

不过这显然还不够，我们从业务出错点直接抛出错误让调用方处理，虽然可行但错误处理显得很零散，业务错误并不是通过类型系统关联到其更高一层（调用方）的错误，错误处理没有体系化。还处于一种指哪打哪的阶段。

此时我们应该考虑将错误处理系统化。

先来看领域驱动设计的六边形架构下的数据流：

![领域驱动设计的六边形架构下的数据流](/assets/images/post_imgs/ddd-error-flow-0.png)

用户发送 DTO 到 HTTP Controller 请求一个操作，后者将 DTO 封装成一个 Command 对象（这里我们就假设是一个create/update/delete的变更操作，因此对应的是 Command 对象，如果是查询操作则对应 Query 对象，这里是什么类型的操作并不影响我们理解）。之后 Command 对象进入 Application Service，接着进入 Domain Service，最后抵达真正处理业务逻辑的地方 Entity，Entity 处理完业务逻辑后，数据反方向流动，从 Entity 到 Domain Service，再到 Application Service，最后经过 HTTP Controller 发送给用户。

这里的几个组件看上去有点抽象，我们将它们映射到银行转账的场景中：

* 用户 -> 用户
* DTO -> 申请的转账金额和收付款账户信息
* HTTP Controller -> HTTP Controller
* Command -> TransferCommand (转账命令对象)
* Application Service -> 账户应用 (Account Application)
* Domain Service -> 转账业务域 (Transfer Service)
* Entity -> 付款账户


